File: Java_Version/3_Longest_Substring_Without_Repeating_Characters.java
/*Author: Bochen (mddboc@foxmail.com)
Last Modified: Tue Apr 10 22:28:45 CST 2018*/

/*Given a string, find the length of the longest substring without repeating characters.

        Examples:

        Given "abcabcbb", the answer is "abc", which the length is 3.

        Given "bbbbb", the answer is "b", with the length of 1.

        Given "pwwkew", the answer is "wke", with the length of 3. Note that the answer must be a substring, "pwke" is a subsequence and not a substring.*/


import java.lang.System;
import java.util.*;
import java.lang.Math;
import java.util.HashMap;


class ListNode
{
    int val;
    ListNode next;

    ListNode(int x)
    {
        val = x;
    }
}



public class Main
{
    public static void main(String[] args)
    {
        String s = "bbbbb";

        Solution solution = new Solution();

        int receive = solution.lengthOfLongestSubstring(s);

        System.out.println("haha");

    }


}


class Solution {
    public int lengthOfLongestSubstring(String s) {

        if (s == null || s.length() == 0) {
            return 0;
        }

        HashMap<Character, Integer> characterIntegerHashMap = new HashMap<Character, Integer>();

        int sLength = s.length();
        int result = 0;
        int startIndex = 0;

        for ( int i = 0; i < sLength; i++ )
        {
            if ( characterIntegerHashMap.containsKey( new Character(s.charAt(i)) ) )
            {
                startIndex = Math.max(
                        characterIntegerHashMap.get( new Character(s.charAt(i)) ) + 1, startIndex);
            }

            characterIntegerHashMap.put(new Character(s.charAt(i)), new Integer(i));
            result = Math.max(result, i - startIndex + 1);

        }

        return result;
    }
}

File: C++_Version/3_Longest_Substring_Without_Repeating_Characters.cpp
/*
  Given a string, find the length of the longest substring without repeating
characters.

Example 1:

Input: "abcabcbb"
Output: 3
Explanation: The answer is "abc", with the length of 3.

Example 2:

Input: "bbbbb"
Output: 1
Explanation: The answer is "b", with the length of 1.

Example 3:

Input: "pwwkew"
Output: 3
Explanation: The answer is "wke", with the length of 3.

  Note that the answer must be a substring, "pwke" is a subsequence and not a
substring.
*/

#include <string>
#include <iostream>

using namespace std;

#define max(x, y) x > y ? x : y;

class Solution {
public:
    int lengthOfLongestSubstring(string s) {

        if (s.length() == 0) {
            return 0;
        }

        int result = 0;
        int startIndex = 0;

        int map[256];
        for (int i = 0; i < 256; i++) {
            map[i] = -1;
        }

        for (int i = 0; i < s.length(); i++) {
            char ch = s[i];
            if (map[ch] != -1) {
                startIndex = max(startIndex, map[ch] + 1);
            }
            result = max(result, i - startIndex + 1);
            map[ch] = i;
        }

        return result;
    }
};

int main(int argc, char **args) {

    string inputs[3] = {"abcabcbb", "bbbbb", "pwwkew"};
    int result[3] = {3, 1, 3};

    Solution solution = Solution();

    for (string input : inputs) {
        cout << solution.lengthOfLongestSubstring(input) << "\n";
    }

    return 0;
}

